### ISSUE TITLE
Data race with table driven tests in go

### ISSUE THREAD

**Issue by kristinapathak** [pulumi/pulumi#8466]
## Hello!

- Vote on this issue by adding a 👍 reaction
- To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already)

## Issue details

My team has begun using pulumi with Go.  When we added table driven unit tests, the tests themselves pass but data races are detected when `go test` is run with the race flag set.  For the specific test in the example, we don't have any values shared between sub tests, but there is still a memory address being accessed by multiple goroutines.

### Steps to reproduce
Link to an example: https://github.com/kristinapathak/pulumi-go-data-race

1. Write some go code that uses pulumi functions.
1. Write a table driven test using go, making sure that no pointers or shared memory addresses are passed to pulumi functions.
1. Run the tests and ensure they pass.
1. Run the tests with the race flag set.

Expected: The unit tests pass without any data races found.
Actual: Data races are found within the tests.  The pulumi calls in the example are `pulumi.RequireSecret()` and `pulumi.Unsecret()` but ultimately the problem seems to be when calling `github.com/pulumi/pulumi/sdk/v3/go/pulumi.(*OutputState).fulfillValue()` and values are assigned to a `*pulumi.OutputState`.

**Comment by joeduffy** [comment]
It does appear based on what I see that `RequireSecret` and `Unsecret` are the culprits. While `fulfillValue` does its mutations under a lock, the former two methods don't. In theory those writes are atomic, but I haven't evaluated the ordering/serializability implications, and so it's probably just safest to acquire the lock as so:

```diff
diff --git a/sdk/go/pulumi/types.go b/sdk/go/pulumi/types.go
index 264d0b700..527d515a7 100644
--- a/sdk/go/pulumi/types.go
+++ b/sdk/go/pulumi/types.go
@@ -494,7 +494,10 @@ func UnsecretWithContext(ctx context.Context, input Output) Output {
 	var x bool
 	o := toOutputWithContext(ctx, input.getState().join, input, &x)
 	// set immediate secretness ahead of resolution/fulfillment
-	o.getState().secret = false
+	s := o.getState()
+	s.mutex.Lock()
+	s.secret = false
+	s.mutex.Unlock()
 	return o
 }
 
@@ -510,7 +513,10 @@ func ToSecretWithContext(ctx context.Context, input interface{}) Output {
 	x := true
 	o := toOutputWithContext(ctx, nil, input, &x)
 	// set immediate secretness ahead of resolution/fufillment
-	o.getState().secret = true
+	s := o.getState()
+	s.mutex.Lock()
+	s.secret = true
+	s.mutex.Unlock()
 	return o
 }
 ```

Separately, we should probably be running a lot more of our tests with the `-race` flag!

**Comment by joeduffy** [comment]
Am curious, what do you think @t0yv0 and @pgavlin, both about mutating shared state without a lock, and also whether we should be doing more `-race` testing. (I'm sure it slows things down, so maybe nightly rather than on every checkin, but probably is a good safeguard either way.)

**Comment by t0yv0** [comment]
I'd be for running an experiment to see the cost of that extra checking and enabling it in some form (PR or master verif).

**Comment by RobbieMcKinstry** [comment]
We recently enabled race detection everywhere. I don't have the bandwidth at this moment, but we should try the listed reproduction and see if this issue is closable.

**Comment by abhinav** [comment]
Thanks for the easy repro, @kristinapathak, and thanks for flagging @RobbieMcKinstry.
I've confirmed that the race was indeed present in the version @kristinapathak linked to.
This specific race was fixed in #11908.

The provided repro repository doesn't have this issue if you run it with the latest pulumi:

```bash
% go get -u github.com/pulumi/pulumi/sdk/v3@latest
go: downloading github.com/pulumi/pulumi/sdk/v3 v3.55.0
go: downloading github.com/pulumi/pulumi v3.55.0+incompatible
go: upgraded github.com/pulumi/pulumi/sdk/v3 v3.17.1 => v3.55.0
go: upgraded github.com/stretchr/testify v1.7.0 => v1.8.1

% go test -race
PASS
ok      example.com/m   0.292s
```

Closing as fixed.
Again, thanks for the easily verifiable repro, @kristinapathak!
